metagov/metagov/core/signals/__init__.py

```diff
@@ -0,0 +1,4 @@
+import django.dispatch
+
+governance_process_updated = django.dispatch.Signal(providing_args=["instance", "status", "outcome", "errors"])
+platform_event_created = django.dispatch.Signal(providing_args=["instance", "event_type", "data", "initiator", "timestamp"])
```

Comment:
@shaunagm [here's an example](https://github.com/mashton/policykit/blob/bf600d205cdfe9a1aa58ede501bfb54f0ffc84b9/policykit/integrations/github/handlers.py#L17-L18) of how the driver would listen for these signals. In the PolicyKit case, it has separate handlers for each supported platform (like the linked github handlers) as well as a generic handler for events ([here](https://github.com/mashton/policykit/blob/metagov_internal_api/policykit/policyengine/handlers.py)) that supports the ability for policies to evaluate from any platform even if PK doesn't have its own support for it (thiso is what we've been calling "generic metagov trigger")

Reply:
One thing that can be a little tricky with Signals is that they can be triggered multiple times from a single event (not sure why this happens, but it does). And this structure you've created here, with a generic handler and then also a custom handler, would seem to trigger multiple handlers on occasion. 

Given these two things, it seems important to either find a way to make sure events are handled only once, or to make clear in the docs that you can't assume that, for a given event, only one function will get triggered / a given function will only get triggered once.

Reply:
By "Signals can be triggered multiple times" do you mean a single signal can actually be emitted twice, or that if you have N handlers the signal will be received N times, or something else? If you mean the former, the only thing I'm seeing is this https://code.djangoproject.com/wiki/Signals#Helppost_saveseemstobeemittedtwiceforeachsave but seems like it is an issue with peoples code, rather than signals actually being duplicated.


For PolicyKit currently, a signal emitted for "a message was posted in Slack" will be received by two handlers (the "generic" handler and the Slack handler) and so there will be two Actions created –– a generic webhook trigger action, and a "slack post message" action. I don't think that's an issue right now, because 1) the action wont be saved if it doesn't match any policies, 2) the generic triggers are a power-user feature, they are more for experimenting with new plugins and actions. This seems like more of a comment about how PolicyKit is implemented, though. 

We definitely need to provide documentation to explain how drivers should handle the signals, but I'm not sure what we can do _on the metagov side_ to make sure events are only handled once.

Reply:
_By "Signals can be triggered multiple times" do you mean a single signal can actually be emitted twice, or that if you have N handlers the signal will be received N times, or something else?_

Confusingly, I meant both. Sorry!

_but seems like it is an issue with peoples code, rather than signals actually being duplicated._

Fair point, but in my experience it is a common problem when working with signals, especially in reusable apps. 

You're right that there's not really a way to prevent the signal from being emitted multiple times, or to make sure it's handled only once, since that's on the Driver to implement. I guess all we can really do is emphasize the issue in the documentation, and encourage Drivers to write handlers/receivers which use `dispatch_uid`.

Reply:
Okay gotcha. I will update PolicyKit to use the dispatch_uid. I created an issue for the gateway django app documentation: https://github.com/metagov/gateway/issues/154